src/main/java/com/google/devtools/build/lib/remote/RemoteActionFileSystem.java

```diff
@@ -84,27 +90,39 @@ public void updateContext(MetadataInjector metadataInjector) {
   }
 
   void injectTree(SpecialArtifact tree, TreeArtifactValue metadata) {
-    checkNotNull(metadataInjector, "metadataInject is null");
-
-    for (Map.Entry<TreeFileArtifact, FileArtifactValue> entry :
-        metadata.getChildValues().entrySet()) {
+    for (Map.Entry<TreeFileArtifact, FileArtifactValue> entry : metadata.getChildValues()
+        .entrySet()) {
       FileArtifactValue childMetadata = entry.getValue();
       if (childMetadata instanceof RemoteFileArtifactValue) {
         TreeFileArtifact child = entry.getKey();
         injectedMetadata.put(child.getExecPath(), (RemoteFileArtifactValue) childMetadata);
       }
     }
 
-    metadataInjector.injectTree(tree, metadata);
+    injectedTreeMetadata.put(tree.getExecPath(), metadata);
   }
 
   void injectFile(ActionInput file, RemoteFileArtifactValue metadata) {
-    checkNotNull(metadataInjector, "metadataInject is null");
-
     injectedMetadata.put(file.getExecPath(), metadata);
+  }
 
-    if (file instanceof Artifact) {
-      metadataInjector.injectFile((Artifact) file, metadata);
+  void flush() {
+    checkNotNull(metadataInjector, "metadataInject is null");
+
+    for (Map.Entry<PathFragment, Artifact> entry : outputMapping.entrySet()) {
```

Comment:
This seems to be a functional change, as previously `injectFile` and `injectTree` would perform the injection even when the artifact was not an output. Is this intentional? (I'm not aware of a reason to inject non-outputs, but that doesn't mean there isn't one :))

Reply:
It is intentional. The difference is, before this change, we could only inject declared spawn outputs (not arbitrary output) which is a super set of declared action outputs. In most cases they should be same (i.e one action generates one spawn). In the cases that declared spawn outputs have more outputs than declared action outputs, most of them are instance of `ActionInput` which don't get injected.

From the perspective of reading the injected metadata, all that matters is metadata for declared outputs. I can't find a reason for allowing inject metadata of non-declared outputs (buy I may be wrong).